keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,731 @@
+result for troubleshooting purpose and with a Network result in a modern
+standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
+controllers.
+
+For each device, if required, the DRA Driver processing the device allocation
+can report the status of it in the `Status.Devices` of the ResourceClaim
+by using the Kubernetes API.
+
+```golang
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+    ...
+
+    // Devices contains the status of each device allocated for this
+    // claim, as reported by the driver. This can include driver-specific
+    // information. Entries are owned by their respective drivers.
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=driver
+    // +listMapKey=device
+    // +listMapKey=pool
+    // +featureGate=DRAResourceClaimDeviceStatus
+    Devices []AllocatedDeviceStatus `json:"devices,omitempty" protobuf:"bytes,4,opt,name=devices"`
+}
+
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+    // Driver specifies the name of the DRA driver whose kubelet
+    // plugin should be invoked to process the allocation once the claim is
+    // needed on a node.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
+    //
+    // +required
+    Driver string `json:"driver" protobuf:"bytes,1,rep,name=driver"`
+
+    // This name together with the driver name and the device name field
+    // identify which device was allocated (`<driver name>/<pool name>/<device name>`).
+    //
+    // Must not be longer than 253 characters and may contain one or more
+    // DNS sub-domains separated by slashes.
+    //
+    // +required
+    Pool string `json:"pool" protobuf:"bytes,2,rep,name=pool"`
+
+    // Device references one device instance via its name in the driver's
+    // resource pool. It must be a DNS label.
+    //
+    // +required
+    Device string `json:"device" protobuf:"bytes,3,rep,name=device"`
+
+    // Conditions contains the latest observation of the device's state.
+    // If the device has been configured according to the class and claim
+    // config references, the `Ready` condition should be True.
+    //
+    // +optional
+    // +listType=atomic
+    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,4,opt,name=conditions"`
+
+    // Data contains arbitrary driver-specific data.
+    //
+    // +optional
+    Data runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`
+
+    // NetworkData contains network-related information specific to the device.
+    //
+    // +optional
+    NetworkData NetworkDeviceData `json:"networkData,omitempty" protobuf:"bytes,6,opt,name=networkData"`
```

Comment:
Question about `NetworkData` vs. `Data`. So let's consider a situation where I've got users using NetworkAttachmentDefinition style specifications, and so they've got fields like:

```
    Name       string      `json:"name"`
    Interface  string      `json:"interface,omitempty"`
    IPs        []string    `json:"ips,omitempty"`
    Mac        string      `json:"mac,omitempty"`
    Mtu        int         `json:"mtu,omitempty"`
    Default    bool        `json:"default,omitempty"`
    DNS        DNS         `json:"dns,omitempty"`
    DeviceInfo *DeviceInfo `json:"device-info,omitempty"`
    Gateway    []string    `json:"gateway,omitempty"`
```

The so, I have this mapping...

```
    Interface  --> InterfaceName
    IPs  --> Addresses
    Mac --> HWAddress
```

Which is 3/x, so the other values would be remapped to... The data field? 

I'm mostly looking for clarity, and trying to figure out standardization (where it makes sense), and also -- extensibility. That is, that with the `*/network-status:` annotation with `net-attach-def`, we've definitely had to slowly expand that struct over time. So I wonder if we should focus on building some standards for the `Data` field if these other fields would go there? Or, could we have something as part of the `NetworkDeviceData` that is somehow extensible, and maybe think about forming a standard for it via Multinetworking WG?

Reply:
If you want to store the rest, yes, the `Data` field. The main idea is that the Data field is 100% driver specific and so there are no guarantees (by the Kubernetes project) with respect to this data structure or API evolution.

The question I have is what data needs to be consumed by other ecosystem projects? That's what should be in a K8s built-in API.

> Or, could we have something as part of the NetworkDeviceData that is somehow extensible, and maybe think about forming a standard for it via Multinetworking WG?

Maybe `NetworkDeviceData` should be a K8s one-of?

Or, to be really flexible, instead of two fields, we have one field `Data []DeviceData`, sort of like we did originally but now an array instead of just a single value:

```go
type AllocatedDeviceStatus struct {
 ...
 Data []DeviceData
}

type DeviceData struct {
  // exactly one must be populated

  // +optional
  Opaque *runtime.RawExtension

  // +optional
  NetworkData *NetworkDeviceData

  // +optional
  SomeFutureThing *MultinetThing
}
```

The sort of hard part is now we can have an aribtrary number of each type, but maybe that's OK.

Reply:
@dougbtv one question, what of those things are configuration option and what of those things need to return a value from the driver?

It look most of those things are configuration options, that already are reported in status, the thing we try to standardize are the things that are reported by the driver that are not set in configuration.

Reply:
The earlier text implies that `Data` may include multiple representations of the same data in different formats:

>     `Data` being an arbitrary data field allows the DRA Driver to store
>     device specific data in different formats. For example, a Network Device being
>     configured by via a CNI plugin could get its `Data` field filled with the CNI
>     result for troubleshooting purpose and with a Network result in a modern
>     standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
>     controllers.

Or maybe that's an artifact from an earlier version of the text and the second part is supposed to be the `NetworkData` field?

If there is any intention that `Data` could have multiple representations of the same data, then it seems like it ought to be explicitly an array or map. Otherwise how are you supposed to be able to reliably extract the representation you wanted?

So I think I like John's version above.

Reply:
yes, this is definitely not for config - config goes in the `DeviceClaim.Config` list, which is a bunch of RawExtensions (or in the `DeviceClass.Config`).

Reply:
Let me put an example on how it will look like (you can embed the json directly if you want in `opaque` but I always hear you complain about it 😄 

User configures network attachment definition in a 
```
apiVersion: resource.k8s.io/v1alpha3
kind:  ResourceClaim
metadata:
  name: multus-test-1
spec:
  devices:
    requests:
      selectors:
        - cel:
            expression: device.driver == "multus.dra.networking.driver"
    config:
    - opaque:
        driver: multus.dra.networking.driver
        parameters:
 	  cniVersion: 0.3.1
  	  type: macvlan
  	  master: ens2
  	  mode: bridge
  	  ipam:
    	  - type: whereabouts
    	     range: 192.168.1.0/24
```

Then it creates a Pod referencing it

```
---
apiVersion: v1
kind: Pod
metadata:
  name: pod1
  labels:
    app: pod
spec:
  containers:
  - name: ctr1
    image: registry.k8s.io/e2e-test-images/agnhost:2.39
  resourceClaims:
  - name: dummy1
    resourceClaimName: multus-test-1
```

The pod gets scheduled and now the ResourceClaim reports back in status

1. The config you have passed
2. The standardize NetworkData: IPs, mac, ... that can be used across the ecosystem to create endpoint slices controller per example
3. The opaque Data object with data that the specific vendor wants to report back from the claim allocation

Reply:
I can see this a number of ways, depending on some answers to above questions.

Do we think that NetworkData is mutually-exclusive with data (i.e. one-of) ?  Or can they both be set at the same time?

"Data" is a hateful word.  I wonder if we can minimize it or find something better.

Option 1: a series of options, not exclusive.
```
    Data *runtime.RawExtension
    NetworkData *DeviceNetworkData
    FrobberData *DeviceFrobberData
```

Option 2: Mutually exclusive, but extgensible options.
```
type DeviceNetworkData struct {
    InterfaceName string
    Addresses []string
    Ext *runtime.RawExtension
}

type DeviceFrobberData struct {
    ID string
    Ext *runtime.RawExtension
}
```

Option 3: One of the above, but nested:
```
type ResourceClaimStatus struct {
  // ...
  Devices []AllocatedDeviceStatus
}

type AllocatedDeviceStatus struct {
  // ...
  Data AllocatedDeviceStatusData
}

type AllocatedDeviceStatusData struct {
    Opaque *runtime.RawExtension
    Network *DeviceNetworkData
    Frobber *DeviceFrobberData
}
```

Or some other arrangement?  We can fine-detail it in API review, but the question is mostly about clarity - which things are extensible, which are fixed, which are composed, and which are exclusive.

Reply:
My suggestion is your 3) except with a slice:`Data []AllocatableDeviceStatusData`. This works for compound device drivers too, or devices that provide multiple functions.

Reply:
> "Data" is a hateful word. I wonder if we can minimize it or find something better.

We poked around for other uses of `RawExtension` and that one came up a few times. Being a bit more thorough:

```console
[hi on] jbelamaric@jbelamaric:~/proj/gh/johnbelamaric/kubernetes/kubernetes$ grep RawExtension $(find . -name types.go) | grep -v // | cut -f 2 -d : | cut -f 1 -d ' ' | tr -d '\t ' | sort | uniq -c | sort -k 1nr
      4 Data
      4 Object
      3 Extension
      2 Config
      2 ConvertedObjects
      2 Items
      2 Objects
      2 OldObject
      2 Options
      2 Parameters
      1 Args
      1 EmptyObject
      1 rawExtension
      1 type
[hi on] jbelamaric@jbelamaric:~/proj/gh/johnbelamaric/kubernetes/kubernetes$ 
```

So...not sure `Object` or `Objects` or `Items` is better than `Data`. And it's not `Config` or `Parameters` or `Extensions` to me. `Info` seems at least as bad as `Data`. Other ideas?

Reply:
It's not mutually exclusive multi-function devices. In particular, I want to build a driver that allocates multiple underlying devices as a single logical unit. Each of those has its own driver with its own output. 

And yes, DriverOutput sounds good to me

Reply:
In context:
   `obj.status.devices[i].data.{opaque,network,whatever}`
   `obj.status.devices[i].driverOutput.{opaque,network,whatever}`
   `obj.status.devices[i].results.{opaque,network,whatever}`
   `obj.status.devices[i].info.{opaque,network,whatever}`

Staring at that, `data` seems not-so-bad.  I don't think `driverOutput` is better.  What about

   `obj.status.devices[i].driver.{opaque,network,whatever}`
or
  `obj.status.devices[i].status.{opaque,network,whatever}`

?

> My suggestion is your 3) except with a slice

So instead of: 

```
type AllocatedDeviceStatus struct {
    // ...

    // At least one of the following must be set. As this evolves
    // over time, back-rev clients might not know about new options,
    // and so might see none of the fields they know.  
    Data *runtime.RawExtension
    NetworkData *DeviceNetworkData
    FrobberData *DeviceFrobberData
}
```
or

```
type AllocatedDeviceStatus struct {
  // ...
  Data AllocatedDeviceStatusData
}

type AllocatedDeviceStatusData struct {
    // At least one of the following must be set. As this evolves
    // over time, back-rev clients might not know about new options,
    // and so might miss some data or see none of the fields they know.
    Opaque *runtime.RawExtension
    Network *DeviceNetworkData
    Frobber *DeviceFrobberData
}
```

You are saying:

```
type AllocatedDeviceStatus struct {
    // ...

    // Each item in this list is expected to have one sub-option set.
    // If a client finds an empty item in this list, they can infer that
    // there was some data that they do not understand.
    Data []AllocatedDeviceStatusData
}

type AllocatedDeviceStatusData struct {
    // Exactly one of the following must be set. As this evolves
    // over time, back-rev clients might not know about new options,
    // and so might see none of the fields they know.  
    Opaque *runtime.RawExtension
    Network *DeviceNetworkData
    Frobber *DeviceFrobberData
}
```

The main difference is that clients can now be aware that they are missing information.

a) We do this in other DRA places, but in this context it seems less critical (status vs spec) -- if the client needed that info, it would know about it.
b) We don't generally handle unknown fields well in clients, and it seems like a systemic solution would be better.

I am not against a list here - it's consistent with other DRA places.  Let's pick a model and name for the KEP and then revisit this detail in code review.

Reply:
I see, I misinterpreted it sorry @dougbtv 

 @johnbelamaric is the list not going to be confusing to identify which information goes where?

Reply:
> @johnbelamaric is the list not going to be confusing to identify which information goes where?

Yes, each list entry needs an identifier if we do this. Maybe make it a list-map?

> `obj.status.devices[i].driver.{opaque,network,whatever}`

`driver` already exists in `device[i]` as the name of the driver.

> `obj.status.devices[i].status.{opaque,network,whatever}`

Meh. Not better than `data` to me.

With data as a slice as I am suggesting:

`obj.status.devices[i].data[j].{opaque,network,whatever}`

Reply:
> `obj.status.devices[i].data[j].{opaque,network,whatever}` 

So a claim has multiple devices, each device has multiple datas (datae?), each of which has exactly one field set, right?

Reply:
yes, that's right. fun stuff

Reply:
`data` is a list-map with some sort of driver-defined identifier for the key